src/applyMiddleware.ts

```diff
@@ -0,0 +1,79 @@
+import compose from './compose'
+import { Middleware, MiddlewareAPI } from './types/middleware'
+import { AnyAction } from './types/actions'
+import { StoreEnhancer, StoreCreator, Dispatch } from './types/store'
+import { Reducer } from './types/reducers'
+
+/**
+ * Creates a store enhancer that applies middleware to the dispatch method
+ * of the Redux store. This is handy for a variety of tasks, such as expressing
+ * asynchronous actions in a concise manner, or logging every action payload.
+ *
+ * See `redux-thunk` package as an example of the Redux middleware.
+ *
+ * Because middleware is potentially asynchronous, this should be the first
+ * store enhancer in the composition chain.
+ *
+ * Note that each middleware will be given the `dispatch` and `getState` functions
+ * as named arguments.
+ *
+ * @param {...Function} middlewares The middleware chain to be applied.
+ * @returns {Function} A store enhancer applying the middleware.
+ */
+export default function applyMiddleware(): StoreEnhancer
+export default function applyMiddleware<Ext1, S>(
+  middleware1: Middleware<Ext1, S, any>
+): StoreEnhancer<{ dispatch: Ext1 }>
+export default function applyMiddleware<Ext1, Ext2, S>(
+  middleware1: Middleware<Ext1, S, any>,
+  middleware2: Middleware<Ext2, S, any>
+): StoreEnhancer<{ dispatch: Ext1 & Ext2 }>
+export default function applyMiddleware<Ext1, Ext2, Ext3, S>(
+  middleware1: Middleware<Ext1, S, any>,
+  middleware2: Middleware<Ext2, S, any>,
+  middleware3: Middleware<Ext3, S, any>
+): StoreEnhancer<{ dispatch: Ext1 & Ext2 & Ext3 }>
+export default function applyMiddleware<Ext1, Ext2, Ext3, Ext4, S>(
+  middleware1: Middleware<Ext1, S, any>,
+  middleware2: Middleware<Ext2, S, any>,
+  middleware3: Middleware<Ext3, S, any>,
+  middleware4: Middleware<Ext4, S, any>
+): StoreEnhancer<{ dispatch: Ext1 & Ext2 & Ext3 & Ext4 }>
+export default function applyMiddleware<Ext1, Ext2, Ext3, Ext4, Ext5, S>(
```

Comment:
Is this the only way in typescript to tell that that is possible to pass an N number of middlewares?

Reply:
Unfortunately, yes, because in this case it's not an array with the same type throughout like a `string[]` or even a `Foo[]`. Middlewares can be a variety of many different types. It's really the `compose` function that's not type friendly.

That said, if you're willing to entertain a breaking change, we can fix this by doing something like [this TS playground](https://www.typescriptlang.org/play/#code/GYVwdgxgLglg9mABBAFgQxmAPAFQHwAUwccAXIjgJSIDeAUIoqJLAogEZoBOB19jjLgFMoILkmJwGiAL7TOXAHSoMSALyIsAVUIBHEAA9yW6mrzJ0mAvoOI0AZwqIAZIhPTho8R2505dCAR7KERhexAAGxCNFSsaO3IAcgBBRNlKZUswAnj2JIAhNJkM2Oz4iCSAYSLKXjowyKhFNEQAelbEAHkAaXqhcKjFdjaOnr6BpogRrt6GwYATaYBRACUVzpW6IA):

```ts
function chain<T>(foo: T) {
  function bar() {
    return foo
  }
  bar.chain = <U>(qux: U) => chain(qux as T & U)
  return bar
}
```
This way, you can apply multiple middlewares like so:

```ts
const result = chain({ a: 'A' }).chain({ b: 'B' }).chain({ c: 'C' })()
result.a // OK
result.b // OK
result.c // OK
result.d // ERROR
```

Reply:
You might also be able to pull off something [like this](https://www.typescriptlang.org/play/#code/JYOwLgpgTgZghgYwgAgCLAM4Ac5gQCwB4AVAPmQG8AoZZAE0xz3wC5kAKRMYAexDeIBKZAF5yABSg8AtpgiEAbj2B1SVAL5UqMAK4gE3Pshg8eJUuwB012XToAbCAHc4UCBgEBtALqC26bFwCc0oaZDcwHSgQSnpGINYOLl5+ZCFRCSlZDAhLNwweewUIdmFNTSoEPgwwZBqeN1FjU3YKZAAjVzYAcgAhAEEAJW7kdQAaWM6ALx6BgC0R9UEterdLBkDmVo6u5D6hxeEAeiPkAHkAaSpV3I2mAm3p2f6F0ePTy+uwBtv4rbaAI46AAePQAigBVAAah2QJ2QAFFBoMzoMqEA):

```ts
interface Dispatch<T> {
  dispatch: (action: T) => Promise<void>
}

function foo<T>(...middlewares: T[]): Dispatch<T> {
  return { dispatch: (action: T) => Promise.resolve() }
}

const store = foo({ bar: 'BAR' }, { baz: 'BAZ' })

store.dispatch({ bar: 'BAR' }) // OK
store.dispatch({ baz: 'BAZ' }) // OK
store.dispatch({ qux: 'QUX' }) // ERROR
```

Reply:
Otherwise, you're going to get into some pretty hairy generics and conditional types like this:

```ts
export default function applyMiddleware<M extends Middleware>(
  ...middlewares: M[]
): M extends Middleware<infer E, infer S, infer D>
  ? StoreEnhancer<E, S> & D
  : never {
  return ((createStore: M extends Middleware<infer E, infer S>
    ? StoreEnhancerStoreCreator<E, S>
    : never) => <
    A extends Action = AnyAction
  >(
    reducer: M extends Middleware<never, infer S> ? Reducer<S, A> : never,
    ...args: any[]
  ) => {
```

Reply:
I think it's easiest (for now) if you just cast the inner return type `as any` so at least the function signature isn't so messy.

```ts
export default function applyMiddleware<T extends Middleware>(
  ...middlewares: T[]
): T extends Middleware<infer E, infer S, infer D>
  ? StoreEnhancer<E, S> & D
  : never {
  return ((createStore: StoreEnhancerStoreCreator) => (
    reducer: any,
    ...args: any[]
  ) => {/* guts */}) as any
}
```

Reply:
Doesn't TS now have some kind of a "mapped types" capability or something like that? I know @nickmccurdy was talking about maybe using that to simplify Reselect's types.

Reply:
Not sure if [this is what](https://mariusschulz.com/blog/mapped-type-modifiers-in-typescript) you are talking about, but I fail to see how that would give us anything here.

Reply:
I'm probably not using the right concept name.  I think Nick was referring to the improved generic type inference in 3.4:

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html

Reply:
check this out: https://github.com/pirix-gh/medium/blob/master/types-curry-ramda/src/bonus.ts

I think @markerikson may be on to something, I found this 3 links deep from the issue referenced in the release notes

Reply:
Nice find! Of particular interest and relevancy is the [`compose`](https://github.com/pirix-gh/medium/blob/294725cd723b1a66cfb8b441c862a0d2d5bd184d/types-curry-ramda/src/bonus.ts#L228-L229) function.

Reply:
it's GPL, so I am trying to learn from it and write new code from memory atm

Reply:
@cellog see #3563

Reply:
@cellog see #3563